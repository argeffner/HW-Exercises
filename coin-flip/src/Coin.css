.Coin-image {
    width: 200px;
}

p {
    font-weight: bold;
}