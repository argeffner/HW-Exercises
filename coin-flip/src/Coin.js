import React, {useState} from "react";
import "./Coin.css";

Coin.defaultProps = {
    coin: [
    {side: 'heads', 
     image: 'https://www.moneymetals.com/images/products/2016-gold-eagles-tenth-obverse.jpg'
    },
    {side: 'tails',
     image: 'https://www.moneymetals.com/images/products/20dollar-st-gaudens-double-eagle-coin-reverse.jpg'
    },
]};

function Coin(props) {
        const [coin, setCoin] = useState(null);
        const [headCount, setHeadCount] = useState(0);
        const [tailCount, setTailCount] = useState(0);

      function handleClick() {
        const reCoin = props.coin[Math.floor(Math.random() * (2))];
        setCoin(reCoin);
        
        if (reCoin.side === 'heads') { setHeadCount(headCount + 1)};
        if (reCoin.side === 'tails') { setTailCount(tailCount + 1)};
      };
    
      let theCoin = coin ? (
        <img className="Coin-image" src={coin.image} alt={coin.side} />
      ) : null;

    return (
      <div className="Coin">
        <h2 className="Coin-title">Heads or Tails?</h2>
        {theCoin}
        <br></br>
        <button onClick={handleClick}>Flip</button>
        <p> 
            Out of {headCount + tailCount} flips, there have been {headCount} heads
            and {tailCount} tails.
        </p> 
      </div>
    );
  }

export default Coin;