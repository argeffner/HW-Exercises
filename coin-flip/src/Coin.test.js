import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Coin from "./Coin";

beforeEach(function() {
    jest
      .spyOn(Math, "random")
      .mockReturnValueOnce(0.25)
      .mockReturnValueOnce(0.75);
  });

// smoke test
it("renders without crashing", function() {
  render(<Coin />);
});

// snapshot of coin
it("matches snapshot", function() {
    const { asFragment } = render(<Coin />);
    expect(asFragment()).toMatchSnapshot();
  });

it("coin doesn't show on loadpage", function() {
  const { queryByTestId } = render(<Coin />);

  // expect no image before clicking button
  expect(queryByTestId("coin")).toBeNull();
});

it("show coin and count for heads", function() {
  const { getByText, queryByAltText } = render(<Coin />);
  const button = getByText("Flip");

  // flips to heads
  fireEvent.click(button);

  //expect first image to show heads
  expect(queryByAltText("heads")).toBeInTheDocument();
  expect(queryByAltText("tails")).not.toBeInTheDocument();
  // expect phrase to have 1 flip and 1 heads
  expect(
    getByText("Out of 1 flips, there have been 1 heads and 0 tails.")
  ).toBeInTheDocument();
})

it("show coin and count for tails", function() {
    const { getByText, queryByAltText } = render(<Coin />);
    const button = getByText("Flip");
  
    // flips to heads then tails
    fireEvent.click(button);
    fireEvent.click(button);
  
    //expect first image to show heads
    expect(queryByAltText("heads")).not.toBeInTheDocument();
    expect(queryByAltText("tails")).toBeInTheDocument();
    // expect phrase to have 2 flip and 1 heads and 1 tails
    expect(
      getByText("Out of 2 flips, there have been 1 heads and 1 tails.")
    ).toBeInTheDocument();
  })

afterEach(function() {
    Math.random.mockRestore();
  });