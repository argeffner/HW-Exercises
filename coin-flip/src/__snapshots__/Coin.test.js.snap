// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Coin"
  >
    <h2
      class="Coin-title"
    >
      Heads or Tails?
    </h2>
    <br />
    <button>
      Flip
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
